compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs

```diff
@@ -135,8 +136,25 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
         // candidate which assumes $0 == int, one that assumes `$0 ==
         // usize`, etc. This spells an ambiguity.
 
-        let mut candidates = self.filter_impls(candidates, stack.obligation);
-
+        let candidates = self.filter_impls(candidates, stack.obligation);
+        let mut candidates = candidates
+            .into_iter()
+            .map(|c| match c {
+                ImplCandidate(impl_def)
+                    if self.tcx().has_attr(impl_def, sym::rustc_select_quick_discard) =>
```

Comment:
My suggestion was roughly this:
```rust
tcx.impl_trait_ref(impl_def).substs.iter().all(|arg| match arg.unpack() {
    GenericArgKind::Lifetime(_) => true,
    GenericArgKind::Type(ty) => matches!(ty::Param(_), ty.kind()),
    GenericArgKind::Const(ct) => matches!(ty::ConstKind::Param(_), ct.kind()),
})
```
(i.e. a "fully blanket" `impl`)

Alternatively, only `Self` could be considered, but "fully blanket" is nicely conservative.